### ISSUE TITLE
Increase output detail of Test-Connection when performing a tcp connection test 

### ISSUE THREAD

**Issue by jackdcasey** [PowerShell/PowerShell#11440]
# Increase output detail of Test-Connection when performing a tcp connection test 

My proposed change is to provide more detail when using `Test-Connection` cmdlet with the `-TCPPort` option. This would bring the command more in line with the classic `Test-NetConnection` cmdlet. 

Currently, the returned value when performing this test is a boolean true / false. This provides some insight into if the connection was successful but not too much else. Here's an example output:

    Test-Connection google.com -TCPPort 443

    True

My proposed changes would provide more detail to the output. Here's an example of what that output might look like:

    Test-Connection google.com -TCPPort 443

    Source             : Jack-PC
    SourceAddress      : 192.168.1.1
    Destination        : google.com
    DestinationAddress : 216.58.217.46
    Port               : 443
    Latency            : 16
    TestResult         : Success

To provide the original behaviour, we can implement the existing -quiet option:

    Test-Connection google.com -TCPPort 443 -Quiet

    True

In my dayjob, I use the classic `Test-NetConnection` cmdlet all the time to troubleshoot network connectivity issues. I find the added output incredibly valuable. I believe this would increase the usability of this command by a fair bit.  

I have forked the repository and have these changes implemented. However, before submitting any pull request I wanted to engage the community, as this is changing an existing implementation. 

Please let me know what you think! I am new to this community, and would love any feedback of any kind :)

**Comment by vexx32** [comment]
Love it!

We should probably look to format the output similarly to how the existing output type are formatted (table format, possibly with multiple connection attempts so that if there's a transient failure it doesn't give the wrong impression).

I would also add a `Mode` column, since from memory the cmdlet also supports `-UdpPort`, and then you could use this some output type for each of them. `Mode` would be a custom enum type where the available values (currently, at least) are `TCP` and `UDP` (or if there's a .NET core enum that has a similar purpose we can always use that).

**Comment by iSazonov** [comment]
> also supports -UdpPort

You can not connect to UDP port. UDP nature is "send and forget".

**Comment by vexx32** [comment]
Ah, my mistake, I was under the impression we'd already done a `-UdpPort`. Feel free to disregard that part. 🙂 

I did look into that at some point, though, since some folx were wanting that. Apparently typical UDP port test tools just do the reverse of the TCP port test -- send a UDP packet, if any response is received, it is assumed that port is closed. Otherwise you have to assume it's open. Naturally that leads to a lot of false positives, but given then nature of UDP that's about as good as you can get. :joy:

**Comment by jackdcasey** [comment]
Thank you for the feedback 🙂!

I was looking at formatting it a bit more like the ping. However, in the current implementation the connection test is only sent once, then the result is returned. Do you reckon we change the logic to perform multiple connectivity tests? 

Let me know what you think!

**Comment by vexx32** [comment]
I think that would be worth doing. It's always better to test a few times in case it hits a transient failure, imo. 🙂 

It makes more sense that we do that for this too, I think.

**Comment by jackdcasey** [comment]
Sounds like a plan! 

I guess for the `-quiet` option we can keep at a single test, as it's only returning a single bool. I'll make the changes to run the test multiple times. 🙂

Pardon my ignorance on the subject, I'm fairly new with contributing to open source. The workflow at my workplace is totally different! Once we are happy with our changes, would we just submit this as a normal pull request? As we're changing an existing cmdlet, are there any other steps required? 

My understanding is it would look like this:

1. Make changes to the command
2. Update tests to reflect changes
3. Update documentation to reflect changes
4. Submit pull request

Please let me know if I am missing anything! 😊 

Thanks!

**Comment by iSazonov** [comment]
> However, in the current implementation the connection test is only sent once, then the result is returned. 

I think follow attempts make sense only on failure.

> Please let me know if I am missing anything! 

Every PR in the repo should has tests to cover code changes.
You could look already merged PRs and our contribution docs.

**Comment by vexx32** [comment]
Sounds good to me @jackdcasey! As Ilya mentions, make sure to have tests that cover it.

As to the repeated tests for TCP with `-Quiet`, I think it makes sense to simply do 4 attempts, and then return `true` the first time you're able to connect. Then you'd only return `false` if all 4 attempts failed to connect, similar to how the ICMP `-Quiet` mode works, if I recall correctly. 🙂

**Comment by iSazonov** [comment]
Really there is two scenario:
- TCP send SYN timeout could be from 20 to 60 seconds (calculated dynamically)
- target application response time

What we want to address?

**Comment by jackdcasey** [comment]
Working on the formatting a bit today, is this what you guys had in mind?
```
Test-Connection google.com -tcpport 443

   Destination: google.com

Test Source           Destination      DestinationAddress    Port Latency Result
                                                                     (ms)
---- ------           -----------      ------------------    ---- ------- ------
   1 Jacks-MBP        google.com       172.217.3.174          443      26 Success
   2 Jacks-MBP        google.com       172.217.3.174          443       9 Success
   3 Jacks-MBP        google.com       172.217.3.174          443      23 Success
   4 Jacks-MBP        google.com       172.217.3.174          443      18 Success
```
In regards to the `-quiet` logic, I reckon we err on the side of caution, and only return true if all 4 tests are successful. As soon as we hit a failed request, we return false. 

Let me know what you guys think!

**Comment by vexx32** [comment]
Looks good to me! 

@iSazonov I'll be first to admit I have limited experience with exactly what can cause a one-off connection error, but I think that since hardware is never perfect we should always attempt a few times to give a reasonably complete picture of how the connection behaves.

**Comment by iSazonov** [comment]
TCP is a connection-based and reliable protocol. To establish a connection the protocol does a conversation which may take a long time, specially if the connection is blocked by firewall. For this reason I'd use default Count = 1.
Count > 1 could be useful for WAIT scenario: "ping" a service until it start and response. This assumes that we stop at first successful connection.
If we want get a latency we should measure from sending first IP to getting last (connection) IP package.
Since establishing TCP connection is OS level, not application, I thing it makes no sense to measure some successful connections.

**Comment by vexx32** [comment]
@iSazonov I suppose I was just thinking it would probably make the most sense for users if both tests behaved somewhat similarly on the surface. But you're right, different protocol, different rules.

So we should make it clear in our documentation that the default for `-TcpPort` tests is one connection, but still allow the use of `-Count $number` or `-Repeat` options for users to make more extensive tests as they need, as well as `-Quiet` if needed.

**Comment by jackdcasey** [comment]
I think this is an awesome idea! It makes sense to have the default behaviour be a single test, with added tests added using the `-count` or `-repeat` option.

I also really like wait scenario you mentioned @iSazonov! In a ton of scenarios we are waiting for services to come online that have icmp totally blocked. Having another reliable option is fantastic!

I'll make those changes soon, it should be fairly straightforward!

**Comment by doctordns** [comment]
one question for @jackdcasey - is the 'latency' a one way measurement, or RTT time? If the latter, would the property be better named 'RTT' or 'RTTTime'?

**Comment by jackdcasey** [comment]
It's measuring the time taken to establish the connection, I reckon measurement of the syn-ack-synack would err on the RTT side. I'd be happy to give a more descriptive label. 

I wasn't super familiar with the acronym RTT, and had to look it up 😂. How would you feel about RoundTripTime as the label?

**Comment by vexx32** [comment]
I'm not opposed to it, but I will mention that having such a long column name doesn't look very nice in the table formats. 🙂 

That's one of the reasons I chose the name `Latency` over `RoundtripTime` (which is, from memory, what the actual underlying property on a `PingReply` object is in .NET Core). I would probably err on the side of consistency there and keep the same name, unless there's a need for a different name here.

**Comment by jackdcasey** [comment]
I agree, keeping it labeled as `latency` may be the best option. Consistency between tests may be nice to have. A nice compromise may be to specify these details in the documentation. What do you guys think about this?

**Comment by jackdcasey** [comment]
I've made the changes mentioned above. By default, the tcp test will only run once, unless specified with `-count` or `-repeat`. Here are some sample outputs:
```
Test-Connection google.com -tcpport 443

   Destination: google.com

Test Source           Destination      DestinationAddress    Port Latency Result
                                                                     (ms)
---- ------           -----------      ------------------    ---- ------- ------
   1 Jacks-MBP        google.com       172.217.3.174          443      15 Success
```
```
Test-Connection google.com -tcpport 443 -count 4

   Destination: google.com

Test Source           Destination      DestinationAddress    Port Latency Result
                                                                     (ms)
---- ------           -----------      ------------------    ---- ------- ------
   1 Jacks-MBP        google.com       172.217.3.174          443      13 Success
   2 Jacks-MBP        google.com       172.217.3.174          443      10 Success
   3 Jacks-MBP        google.com       172.217.3.174          443      16 Success
   4 Jacks-MBP        google.com       172.217.3.174          443      16 Success
```
```
Test-Connection google.com -tcpport 443 -repeat 

   Destination: google.com

Test Source           Destination      DestinationAddress    Port Latency Result
                                                                     (ms)
---- ------           -----------      ------------------    ---- ------- ------
   1 Jacks-MBP        google.com       172.217.3.174          443      15 Success
   2 Jacks-MBP        google.com       172.217.3.174          443      15 Success
   3 Jacks-MBP        google.com       172.217.3.174          443      10 Success
   4 Jacks-MBP        google.com       172.217.3.174          443      16 Success
   5 Jacks-MBP        google.com       172.217.3.174          443      16 Success
   6 Jacks-MBP        google.com       172.217.3.174          443      15 Success
   7 Jacks-MBP        google.com       172.217.3.174          443      18 Success
   8 Jacks-MBP        google.com       172.217.3.174          443      16 Success
   9 Jacks-MBP        google.com       172.217.3.174          443      16 Success
  10 Jacks-MBP        google.com       172.217.3.174          443      17 Success
   ...
```
Let me know if this matches what you guys had in mind!

**Comment by jackdcasey** [comment]
@iSazonov To play with the wait scenario you mentioned earlier, I ran a `-repeat` tcp test against a web server, while rebooting it. The results were pretty neat! 
```
Test-Connection jacktest.com -tcpport 443 -repeat        

   Destination: jacktest.com

Test Source           Destination      DestinationAddress    Port Latency Result
                                                                     (ms)
---- ------           -----------      ------------------    ---- ------- ------

~ omitted ~

  35 Jacks-MBP        jacktest.com     35.233.210.37          443      15 Success
  36 Jacks-MBP        jacktest.com     35.233.210.37          443      16 Success
  37 Jacks-MBP        jacktest.com     35.233.210.37          443      23 Success
  38 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  39 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  40 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  41 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Timeout
  42 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Timeout
  43 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Timeout
  44 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Timeout
  45 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  46 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  47 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  48 Jacks-MBP        jacktest.com     35.233.210.37          443       0 Failed
  49 Jacks-MBP        jacktest.com     35.233.210.37          443      23 Success
  50 Jacks-MBP        jacktest.com     35.233.210.37          443      21 Success
  51 Jacks-MBP        jacktest.com     35.233.210.37          443      21 Success
  52 Jacks-MBP        jacktest.com     35.233.210.37          443      15 Success
```
I will definitely use this in the future when waiting for a server to come up!

**Comment by iSazonov** [comment]
@jackdcasey I'd prefer to keep a PR as small as possible for fast review and merge. So please start with implementing only "default Count = 1 for TCPPort". In follow PRs we could address other scenarios.

**Comment by vexx32** [comment]
@iSazonov I have my doubts that any of this will make it in for 7.0 GA at this point in time... but yeah, keeping it short and sweet will give it the best shot.

@jackdcasey looks awesome, nice work!

**Comment by iSazonov** [comment]
> I have my doubts

Later this will be a breaking change _after_ 7.0 release but now - before 7.0.

**Comment by jackdcasey** [comment]
I see what you mean! I'll try to boil things down to just the necessary changes so far. The biggest change was the output formatting, most of the other features were already part of the command with some small tweaks.

I'll get that done, and submit the request, unless you guys have any other suggestions 😁

Appreciate all the insight!

**Comment by doctordns** [comment]
> It's measuring the time taken to establish the connection, I reckon measurement of the syn-ack-synack would err on the RTT side. I'd be happy to give a more descriptive label.
> 
> I wasn't super familiar with the acronym RTT, and had to look it up 😂. How would you feel about RoundTripTime as the label?

Seems to me that anyone serious about networks and using Test-Connection probably knows what RTT is! :-) ANd it's fewer characters!

**Comment by vexx32** [comment]
@doctordns you're not wrong, but in general I think it best to avoid abbreviations there. Abbreviations tend to be hard to Google for those who don't already know them. ^^

**Comment by copdips** [comment]
@jackdcasey 

I cannot wait to see this can be released, love it.
Just a little suggestion, maybe in the future, people will add the feature to test the UDP port. so the column name `Port` in the output might be not precise enough.
The parameter name is `-TcpPort`, why not use the same name ?
Or two columns: `Port` and `Protocol`

**Comment by jackdcasey** [comment]
@copdips I think the ability to test udp ports would be awesome! 

However as udp is a bit more fire-and-forget compared with tcp, this may be difficult to implement. I've done a bit of research, many methods I've found used to test UDP ports are a bit less reliable. 

I'm trying to keep the scope on this one as minimal as possible, however would be more than excited to play with that idea in the future! 😁

**Comment by doctordns** [comment]
@jackdcasey any sort of UDP test would be near impossible. If an app sends a server a UDP datagram, many (most?) services do not return a response.  And the few that do return a response are not consistent. 

You could hard code a few special cases (eg DNS, NFS. and maybe a few more) where you know what the service would do if you sent it an appropriate UDP 'command'. You could even add a -**CommonUDPPort** parameter (to mirror the **-CommonTCPPort** parameter in **Test-NetConnection**), and create specific tests for DNS, etc.

With DNS, for example, send the target server a query for foo.local, and time the NXDomain response. Report back how long the return took.  You could do something similar for NTP (Network Time Protocol), BOOTP/DHCP. NNP (Network News Protocol), and maybe even Quote of the day protocol., TFTP, RTSP, RIP, and OSPF. It's been a while since I studied these protocols, but some of these could be tested/measured using carefully constructed UDP packets that return a specific expected response. 

Not sure it's worth it, but testing UDP services like this would be very cool. Testing UDP has never been easy! :-)

**Comment by jackdcasey** [comment]
Hey guys, just commenting here too see if there's still any interest in this! 

I have PR #11452 still open, and haven't been able to get much traction 😄

**Comment by iSazonov** [comment]
We expect 7.1 release in weeks and after that I hope MSFT team will have a time to review community PRs.

**Comment by doctordns** [comment]
I'd love to see this PR make it into 7.2, if only as an wxoerime feature.
